silq/tools/circuit_tools.py

```diff
@@ -0,0 +1,175 @@
+        >>> convert_circuit('GxGi', target_type=list)
+        ['Gx', 'Gi']
+    """
+    from pygsti.objects.circuit import Circuit
+    
+    # First convert all types to string
+    if isinstance(circuit, Circuit):
+        expanded_circuit = []
+        for gate in circuit.get_labels():
+            if gate.issimple():
+                expanded_circuit.append(gate.name)
+            elif str(gate) == '[]':
+                gate.append('Gi')
+            else:
+                raise RuntimeError(f'Cannot understand gate {gate}')
+
+    elif isinstance(circuit, str):
+        # Replace [] by identity gate
+        expanded_circuit = circuit.replace('[]', 'Gi')
+
+        # Expand any expressions ({gates})^exponent
+        expand_subcircuit = lambda match: match.group(0).replace(
+            match.group(0), match.group(1) * int(match.group(2))
+        )
+        expanded_circuit = re.sub(r'\(([A-Za-z]+)\)\^([0-9]+)', expand_subcircuit, expanded_circuit)
+
+        # Expand any expressions {gate}^exponent (notice missing parentheses
+        expand_subcircuit = lambda match: match.group(0).replace(
+            match.group(0), ('G'+match.group(1)) * int(match.group(2))
+        )
+        expanded_circuit = re.sub(r'G([a-z0-9:]+)\^([0-9]+)', expand_subcircuit, expanded_circuit)
+
+        # Remove all single parentheses without exponent
+        expanded_circuit = re.sub(r"\(|\)", "", expanded_circuit)
+
+        # Newer circuits end with @(qubit idxs), which should be removed
+        expanded_circuit = expanded_circuit.split('@')[0]
+
+        if expanded_circuit == '{}':
+            expanded_circuit = ''
+
+        # Split by every letter G
+        expanded_circuit = ['G'+ gate for gate in expanded_circuit.split('G')[1:]]
+    elif isinstance(circuit, list):
+        expanded_circuit = circuit
+    else:
+        raise RuntimeError("circuit type not understood")
+
+    if target_type == str:
+        if expanded_circuit:
+            return ''.join(expanded_circuit)
+        else:
+            return '{}'
+    elif target_type == list:
+        return expanded_circuit
+    else:
+        raise NameError(f'Cannot select target type {target_type}')
+
+
+def save_circuits(circuits, filepath):
+    """Save list of circuits to a .txt file"""
+    filepath = Path(filepath)
+    if not filepath.suffix:
+        filepath = filepath.with_suffix('.txt')
+
+    with filepath.open('w') as f:
+        for circuit in circuits:
+            circuit_str = convert_circuit(circuit, target_type=str)
+            f.write(circuit_str + '\n')
+
+
+def load_circuits(
+        filepath: Union[str, Path],
+        target_type: type = list,
+        load_probabilities: bool = False
+):
+    filepath = Path(filepath)
+    with open(filepath.with_suffix('.txt')) as file:
+        lines = [line.strip() for line in file]
+
+    # Check if line already contains data
+    if lines[0].startswith('##'):
+        # Remove two headers
+        lines = lines[1:]
```

Comment:
Looks like you're removing a single header here.

Reply:
fixed